### PR TITLE
Feature/col 544 azure

### PR DESCRIPTION
**Pull request by ayang99**
implementation of upload/download against an Azure Blob Storage repo

### CODE REVIEW

**Comment by btiernay** [comment]
Ok to test

**Comment by btiernay** [comment]
Test this please

**Comment by btiernay** [comment]
ok to test

**Comment by btiernay** [comment]
ok to test

**Comment by btiernay** [comment]
ok to test

**Comment by btiernay** [comment]
ok to test

**Comment by btiernay** [comment]
Ok to test

**Comment by btiernay** [comment]
ok to test

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadService.java [hunk 1]
```diff
@@ -298,7 +301,16 @@ private void doMd5Checksum(DownloadRequest req, ObjectSpecification spec) {
     }
     val outputFile = req.getOutputFilePath();
     val downloadedMd5 = calculateChecksum(outputFile);
-    val check = downloadedMd5.equals(spec.getObjectMd5());
+    // val check = downloadedMd5.equals(spec.getObjectMd5());
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/download/DownloadService.java]: Please remove.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/config/S3Config.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.context.annotation.Profile;
+
+@Configuration
+@Profile({ "aws", "collab" })
+public class S3Config {
+
+  @Bean
+  public UploadService uploadService() {
+    return new S3UploadService();
+  }
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/config/S3Config.java]: Newline before class closing.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/config/AzureConfig.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.context.annotation.Profile;
+
+@Configuration
+@Profile({ "azure" })
+public class AzureConfig {
+
+  @Bean
+  public UploadService uploadService() {
+    return new AzureUploadService();
+  }
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/config/AzureConfig.java]: Newline before class closing.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    // Calculate expected number of parts to track progress against
+    Pair<Integer, Long> partInfo = this.calculateNumBlocks(file.length(), BLOCK_SIZE);
+
+    // Will contain one Part as far as Storage Service is concerned
+    if (spec.getParts().isEmpty()) {
+      throw new NotRetryableException(new Exception("Did not get an Upload Part definition from service"));
+    }
+
+    // get SAS URL from Part
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: Upper

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    int totalParts = parts.size();
+
+    // Remove completed parts if don't require checksum-ing
+    if (!checksum) {
+      parts.removeIf(new Predicate<Part>() {
+
+        @Override
+        public boolean test(Part part) {
+          return part.isCompleted();
+        }
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: New line

File: dcc-storage-core/src/main/java/org/icgc/dcc/storage/core/util/MD5s.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.codec.binary.Hex;
+
+/*
+ * Using Commons Codec because Java's Integer.parseInt() wasn't behaving as expected. Codec also already a transitive dependency.
```
Comment (btiernay) [inline dcc-storage-core/src/main/java/org/icgc/dcc/storage/core/util/MD5s.java]: What was wrong with Guava?

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureUploadService.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+      return blob.exists();
+    } catch (URISyntaxException e) {
+      throw new NotRetryableException(e);
+    } catch (StorageException e) {
+      throw new NotRetryableException(e);
+    }
+  }
+
+  @Override
+  public void finalizeUploadPart(String objectId, String uploadId, int partNumber, String md5, String eTag) {
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureUploadService.java]: Agreed

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureUploadService.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  public PartCalculator partCalculator;
+
+  @Override
+  public ObjectSpecification initiateUpload(String objectId, long fileSize, String md5, boolean overwrite) {
+    ObjectSpecification result;
+    try {
+      val blob = container.getBlockBlobReference(objectId);
+
+      if (!overwrite) {
+        if (blob.exists()) {
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureUploadService.java]: Agreed

**Comment by btiernay** [comment]
ok to test

**Comment by btiernay** [comment]
Please address the Codacy feedback as well, less the ones about underscores in test method names

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+      val blob = new CloudBlockBlob(sasUrl);
+      val progress = new Progress(terminal, false, Math.toIntExact(partInfo.getLeft()), 0);
+      OperationContext ctx = new OperationContext();
+
+      // Keep track of number of parts completed - we're also going to get an event for the finalize block list call
+      AtomicInteger completedParts = new AtomicInteger();
+
+      // Add listener for every completed call to Azure Blob Storage
+      // This is how we track progress
+      ctx.getResponseReceivedEventHandler().addListener(new StorageEvent<ResponseReceivedEvent>() {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: Lambda?

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+    // Will contain one Part as far as Storage Service is concerned
+    if (spec.getParts().isEmpty()) {
+      throw new NotRetryableException(new Exception("Did not get an Upload Part definition from service"));
+    }
+
+    // Get SAS URL from Part
+    URI sasUrl;
+    try {
+      sasUrl = new URI(spec.getParts().get(0).getUrl());
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: Helpler

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  // This is Azure's maximum 4 MB upload part size
+  private final static int BLOCK_SIZE = 4 * 1024 * 1024;
+
+  @Override
+  public void upload(File file, String objectId, String md5, boolean redo) throws IOException {
+    // Get object specification from server
+    val spec = storageService.initiateUpload(objectId, file.length(), redo, md5);
+
+    // Calculate expected number of parts to track progress against
+    Pair<Integer, Long> partInfo = this.calculateNumBlocks(file.length(), BLOCK_SIZE);
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: `val`

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+      // Add listener for every retry attempt - just for informational purposes
+      ctx.getRetryingEventHandler().addListener(new StorageEvent<RetryingEvent>() {
+
+        @Override
+        public void eventOccurred(RetryingEvent eventArg) {
+          log.info("Upload required a retry - just saying");
+        }
+      });
+
+      BlobRequestOptions options = new BlobRequestOptions();
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: `val` to reduce type noise.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+import com.microsoft.azure.storage.RetryingEvent;
+import com.microsoft.azure.storage.StorageEvent;
+import com.microsoft.azure.storage.StorageException;
+import com.microsoft.azure.storage.blob.BlobRequestOptions;
+import com.microsoft.azure.storage.blob.CloudBlockBlob;
+
+@Slf4j
+public class AzureUploadService implements UploadService {
+
+  @Autowired
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: http://docs.icgc.org/software/development/#field-comments

File: dcc-storage-core/pom.xml [hunk 14]
```diff
@@ -50,6 +50,11 @@ ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF S
       <version>2.6</version>
     </dependency>
     <dependency>
+      <groupId>commons-codec</groupId>
```
Comment (btiernay) [inline dcc-storage-core/pom.xml]: Is this still needed?

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureUploadService.java [hunk 15]
```diff
@@ -106,51 +106,65 @@ public boolean exists(String objectId) {
...
+    // Empty implementation - not applicable for Azure Upload since we're using the Upload implementation supplied by
+    // SDK
   }
 
   @Override
   public List<MultipartUpload> listUploads() {
+    // Empty implementation - not applicable for Azure Upload since we're using the Upload implementation supplied by
+    // SDK
     return Collections.<MultipartUpload> emptyList();
   }
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/repository/azure/AzureUploadService.java]: New line.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+    uploadParts(parts, file, uploadProgress.getObjectId(), uploadProgress.getUploadId(), progress);
+    cleanupState(file, objectId);
+  }
+
+  /**
+   * Calculate the number of completed parts
+   */
+  private int numCompletedParts(List<Part> parts) {
+    int completedTotal = 0;
+    for (Part part : parts) {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: `val`

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+  @SneakyThrows
+  private void uploadParts(List<Part> parts, File file, String objectId, String uploadId, Progress progressBar) {
+    transportBuilder.withProxy(storageService)
+        .withProgressBar(progressBar)
+        .withParts(parts)
+        .withObjectId(objectId)
+        .withTransportMode(Mode.UPLOAD)
+        .withSessionId(uploadId);
+    transportBuilder.build().send(file);
+  }
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]:
```java
    val transport = transportBuilder
        .withProxy(storageService)
        .withProgressBar(progressBar)
        .withParts(parts)
        .withObjectId(objectId)
        .withTransportMode(Mode.UPLOAD)
        .withSessionId(uploadId) .build();
        
    transport.send(file);
```

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+            progress.incrementParts(1);
+            progress.incrementBytesWritten(partCount == partInfo.getLeft() ? bytesSent = partInfo.getRight() : bytesSent);
+          }
+          // Don't increment if all parts have been accounted for (this is the Put Block List call)
+          // This would result in progress bar reporting "Parts: 42/41"
+        }
+      });
+
+      // Add listener for every retry attempt - just for informational purposes
+      ctx.getRetryingEventHandler().addListener(new StorageEvent<RetryingEvent>() {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: Lamda

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+   * The only public method for client to call to upload data to remote storage
+   * 
+   * @param file The file to be uploaded
+   * @param objectId The object id that is used to associate the file in the remote storage
+   * @param redo If redo the upload is required
+   * @throws IOException
+   */
+  @Override
+  public void upload(File file, String objectId, String md5, final boolean redo) throws IOException {
+    boolean flag = redo;
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: We could probably name `flag` better to represent what .

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+          startUpload(file, objectId, md5, flag);
+        } else {
+          // only perform checksum the first time of the resume
+          resumeIfPossible(file, objectId, md5, retry == 0 ? true : false);
+        }
+        return;
+      } catch (NotRetryableException e) {
+        log.warn(
+            "Upload was not completed successfully in the last execution. Checking data integrity. Please wait...");
+        flag = !storageService.isUploadDataRecoverable(objectId, file.length());
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: This assignments appears useless. Please explain its value.

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+    cleanupState(file, objectId);
+  }
+
+  /**
+   * Resume a upload if it is possible. Otherwise, it will start a new upload. Resume might not be possible if the
+   * upload progress cannot be retrieved.
+   */
+  @SneakyThrows
+  private void resumeIfPossible(File uploadFile, String objectId, String md5, boolean checksum) {
+    UploadProgress progress = null;
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: Move this in the `try`

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+        throw new NotResumableException(new IllegalStateException(msg));
+      }
+      // Then local upload id not present
+    } else if (progress != null) {
+      val msg =
+          String
+              .format(
+                  "There is already an upload in progress that was started elsewhere with upload id %s. Aborting local upload.",
+                  progress.getUploadId());
+      throw new NotResumableException(new IllegalStateException(msg));
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: This pattern happens very very often throughout the code base and takes significant amount of space to articulate. I think adding a `checkStateResumable` might be in order. For example:

```java
...
} else {
  checkStateResumable(progress == null, "There is already an upload in progress that was started elsewhere   with upload id %s. Aborting local upload.",  
  progress.getUploadId());
}
```

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+  }
+
+  /**
+   * Resume an upload given the upload progress. Checksum is required only for the first attempt for each process
+   * execution.
+   */
+  private void resume(File file, UploadProgress uploadProgress, String objectId, boolean checksum) throws IOException {
+    log.info("Resume from the previous upload...");
+
+    List<Part> parts = uploadProgress.getParts();
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: `val`

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+    int completedParts = numCompletedParts(parts);
+    int totalParts = parts.size();
+
+    // Remove completed parts if don't require checksum-ing
+    if (!checksum) {
+      parts.removeIf(new Predicate<Part>() {
```
Comment (btiernay) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java]: Lambda!

File: dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/config/AzureServerConfig.java [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+  @Bean
+  public URLGenerator url() {
+    return new AzureURLGenerator();
+  }
+
+  @Bean
+  public BucketNamingService bucketNamingService() {
+    return new AzureBucketNamingService();
+  }
```
Comment (btiernay) [inline dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/config/AzureServerConfig.java]: New line before class closing

File: dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java (hunk 10 above)
Comment (ayang99) [inline dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/azure/AzureUploadService.java]: Comment in line 69

**Pull request merged**